python/mlx/nn/layers/linear.py

```diff
@@ -15,31 +31,98 @@ class Linear(Module):
         bias (bool, optional): If set to ``False`` then the layer will
-          not use a bias. Default ``True``.
+          not use a bias. Default is ``True``.
     """
 
-    def __init__(self, input_dims: int, output_dims: int, bias: bool = True):
+    def __init__(self, input_dims: int, output_dims: int, bias: bool = True) -> None:
         super().__init__()
-        scale = math.sqrt(1 / input_dims)
+        scale = math.sqrt(1.0 / input_dims)
         self.weight = mx.random.uniform(
             low=-scale,
             high=scale,
             shape=(output_dims, input_dims),
         )
         if bias:
-            self.bias = mx.zeros((output_dims,))
+            self.bias = mx.random.uniform(
+                low=-scale,
+                high=scale,
+                shape=(output_dims,),
+            )
 
-    def _extra_repr(self):
+    def _extra_repr(self) -> str:
         return f"input_dims={self.weight.shape[1]}, output_dims={self.weight.shape[0]}, bias={'bias' in self}"
 
-    def __call__(self, x):
+    def __call__(self, x: mx.array) -> mx.array:
         x = x @ self.weight.T
         if "bias" in self:
             x = x + self.bias
         return x
+
+
+class Bilinear(Module):
+    r"""Applies a bilinear transformation to the inputs.
+
+    Concretely:
+
+    .. math::
+
+        y_i = x_1^\top W_i x_2 + b_i
+
+    where
+    :math:`W` has shape ``[output_dims, input1_dims, input2_dims]``.
+    :math:`b` has shape ``[output_dims, ]``.
+    :math:`i` is the index for output dimensions.
+
+    The values are initialized from :math:`\mathcal{U}(-{k}, {k})`, where
+    :math:`k = \frac{1}{\sqrt{input1\_dims}}`.
+
+    Args:
+        input1_dims (int): The dimensionality of the input1 features
+        input2_dims (int): The dimensionality of the input2 features
+        output_dims (int): The dimensionality of the output features
+        bias (bool, optional): If set to ``False`` then the layer will
+          not use a bias. Default is ``True``.
+    """
+
+    def __init__(
+        self, input1_dims: int, input2_dims: int, output_dims: int, bias: bool = True
+    ) -> None:
+        super().__init__()
+        scale = math.sqrt(1.0 / input1_dims)
+        self.weight = mx.random.uniform(
+            low=-scale,
+            high=scale,
+            shape=(1, output_dims, input1_dims, input2_dims),
+        )
+        if bias:
+            self.bias = mx.random.uniform(
+                low=-scale,
+                high=scale,
+                shape=(output_dims,),
+            )
+
+    def _extra_repr(self) -> str:
+        return (
+            f"input1_dims={self.weight.shape[2]}, input2_dims={self.weight.shape[3]}, "
+            f"output_dims={self.weight.shape[1]}, bias={'bias' in self}"
+        )
+
+    def __call__(self, x1: mx.array, x2: mx.array) -> mx.array:
+        x1 = x1.reshape(x1.shape[0], 1, 1, x1.shape[1])
+        x2 = x2.reshape(x2.shape[0], 1, x2.shape[1])
```

Comment:
Could you use `mx.expand_dims` here and negative indexing so that we don't constrain the inputs to be 2 dimensional.

Reply:
That's smart, I've changed it to:
`x1 = mx.expand_dims(x1, axis=(-2, -3))`
`x2 = mx.expand_dims(x2, axis=(-2))`